### DISCUSSION TITLE
Should retrieving an interface from another interface be allowed?

### DISCUSSION THREAD

**Discussion by albeanth** [terrapower/armi#1491] [General]
There was a bit of a discussion on this and I'm curious to know what others think. 

These methods are used fairly prolifically in ARMI and downstream ARMI applications. 
https://github.com/terrapower/armi/blob/f5681d7162b100fa74e290f9a0f7c5e807076a54/armi/operators/operator.py#L908
https://github.com/terrapower/armi/blob/f5681d7162b100fa74e290f9a0f7c5e807076a54/armi/interfaces.py#L421-L422

Is it ok to do so? There's certainly validity to the potential of these methods getting abused. E.g., a specific interface stack order is set and then a plugin dev decides they need an interface, do stuff with that interface, and change the state out of the prescribed order. On the other hand, sometimes functionality from an interface is legitimately needed. Should an import be used instead? Maybe doing so introduces circular imports and other not-fun dependencies. Curious to know what the community thinks here.

**Comment by keckler** [comment]
My opinion is that it should be allowed. It is very very powerful and avoids the proliferation of tons of helper functions being defined everywhere that this is currently being utilized. It also helps with swapping out different physics kernels because you can grab an interface by it's `function`, without having to know the name of it or writing a factory to get it.

**Reply by john-science** [reply]
This makes me nervous, as I have said before. It seems like the kind of thing that could lead to circular code loops and spaghetti code.

I guess we can continue to support this, but I feel we might want to add some sort of large warning that Interfaces usually depend heavily on when they are called, and the current state of the system. For instance, doing what you suggest, you can call the `InterfactBOL()` method at some random time node.  And the original writer of that Interface can't have predicted that or kept the code safe in that situation.

In fact, I very much do NOT want everyone writing Interfaces to have to put a ton of protective "if" statements and "try"/"except" statements everywhere in their code, to protect people using their Interface in a way it wasn't intended.  That will end up being just as bad as spaghetti code.

But, sure, if this is really important, we can technically still support it.
